text/3301-atomic-memcpy.md

```diff
@@ -0,0 +1,227 @@
+    pub fn load_to(&self, dest: &mut MaybeUninit<T>, ordering: Ordering);
+    pub fn store_from(&self, src: &MaybeUninit<T>, ordering: Ordering);
+
+    pub fn load_to_slice(this: &[Self], dest: &mut [MaybeUninit<T>], ordering: Ordering);
+    pub fn store_from_slice(this: &[Self], src: &[MaybeUninit<T>], ordering: Ordering);
+
+    pub const fn into_inner(self) -> MaybeUninit<T>;
+
+    pub const fn as_ptr(&self) -> *const T;
+    pub const fn as_mut_ptr(&self) -> *mut T;
+
+    pub const fn get_mut(&mut self) -> &mut MaybeUninit<T>;
+    pub const fn get_mut_slice(this: &mut [Self]) -> &mut [MaybeUninit<T>];
+
+    pub const fn from_mut(value: &mut MaybeUninit<T>) -> &mut Self;
+    pub const fn from_mut_slice(slice: &mut [MaybeUninit<T>]) -> &mut [Self];
+}
+
+impl<T> Debug for AtomicPerByte<T>;
+impl<T> From<T> for AtomicPerByte<T>;
+impl<T> From<MaybeUninit<T>> for AtomicPerByte<T>;
+```
+
+Note how the entire interface is safe.
+All potential unsafety is captured by using `MaybeUninit`.
+
+The load functions panic if the `ordering` is not `Relaxed` or `Acquire`.
+The store functions panic if the `ordering` is not `Relaxed` or `Release`.
+The slice functions panic if the slices are not of the same length.
+
+# Drawbacks
+
+- In order for this to be efficient, we need an additional intrinsic hooking into
+  special support in LLVM. (Which LLVM needs to have anyway for C++.)
+
+- It's not immediately obvious this type behaves like a `MaybeUninit`,
+  making it easy to forget to manually drop any values that implement `Drop`.
+
+  This could be solved by requiring `T: Copy`, or by using a better name for this type.
+
+- `MaybeUninit<T>` today isn't as ergonomic as it should be.
+
+  For a simple `Copy` type like `u8` it might be nicer to be able to use types like `&[u8]`
+  rather than `&[MaybeUninit<u8>]`, etc.
+  (But that's a larger problem affecting many other things, like `MaybeUninit`'s interface,
+  `Read::read_buf`, etc. Maybe this should be solved separately.)
+
+# Alternatives
+
+- Instead of a type, this could all be just two functions on raw pointers,
+  such as something like `std::ptr::copy_nonoverlaping_load_atomic_per_byte`.
+
+  This means having to use `UnsafeCell` and more unsafe code wherever this functionality is used.
+
+  It'd be inconsistent with the other atomic operations.
+  We don't have e.g. `std::ptr::load_atomic` that operates on pointers either.
+
+- Require `T: Copy` for `AtomicPerByte<T>`, such that we don't need to worry about
+  duplicating non-`Copy` data.
+
+  There might be valid use cases with non-`Copy` data.
+  Also, not all "memcpy'able" data is always marked as `Copy` (e.g. to prevent implicit copies).
+
+- Leave this to the ecosystem, outside of the standard library.
+
+  Since this requires special compiler support, a community crate would
+  have to use (probably technically unsound) hacks like volatile operations
+  or inline assembly.
+
+# Unresolved questions
+
+- Should we require `T: Copy`?
```

Comment:
Given this is a quite advanced tool, I would lean towards not requiring it.

Some documentation in `AtomicPerByte` could ameliorate the issue.

Reply:
Yes, IMO this should not require Copy, since there's no way to shake off this requirement. I think it's reasonable to use this for a `T` which happens to hold something else in an UnsafeCell in it, which would make it non-Copy.